internal/command/test.go

```diff
@@ -1065,9 +1066,31 @@ func getEvalContextFromStates(states map[string]*TestFileState) *hcl.EvalContext
 		}
 		runCtx[state.Run.Name] = cty.ObjectVal(outputs)
 	}
-	ctx := &hcl.EvalContext{Variables: map[string]cty.Value{"run": cty.ObjectVal(runCtx)}}
 
-	return ctx
+	varCtx := make(map[string]cty.Value)
+	for _, variable := range config.Module.Variables {
+		// If the variable is referenced in the tfvars file or TF_VAR_ environment variable, then lookup the value
+		// in global variables; otherwise, assign the default value.
+		if globals != nil {
+			if vv, exists := globals[variable.Name]; exists {
+				val, valDiags := vv.ParseVariableValue(variable.ParsingMode)
+				if valDiags == nil {
+					varCtx[variable.Name] = val.Value
+					continue
+				}
+				diags = diags.Append(valDiags)
+			}
+		}
+		varCtx[variable.Name] = variable.Default
+	}
```

Comment:
Could we use an abstraction like `prepareInputVariablesForAssertions` here? Looks like it does something very similar (populating variables and default values for test assertions)

It seems that you could extract a chunk of the `prepareInputVariablesForAssertions` code and use it here, instead of reimplementing it here

Reply:
I have made the requested changes in latest [commit.](https://github.com/opentofu/opentofu/pull/1488/commits/0777ea9d965f9269039da0ad55654637576eb30d)